### ISSUE TITLE
Implement "fold" node in MX

### ISSUE THREAD

**Issue by jaeandersson** [casadi/casadi#1395]
Cf. #1064.

**Comment by ghorn** [comment]
I also want `mapAccum`. Copy-paste from previous thread:

Prototypes:
```c++
vector<MX> Function::fold(vector<MX>, vector<vector<MX>>);

pair<vector<vector<MX>>, vector<MX>> Function::mapAccum(vector<MX>, vector<vector<MX>>);
```

expanded fold
```c++
vector<MX> Function::fold(vector<MX> &acc0, vector<vector<MX>> &inputs)
    vector<MX> acc = acc0;
    for (int k=0; k<inputs.size(); k++){
        vector<MX> local_inputs = {acc, inputs[k]}; // psuedo-code
        acc = this.call(local_inputs);
    }
    return acc;
}
```

expanded mapAccum
```c++
pair<vector<vector<MX>>, vector<MX>> Function::mapAccum(vector<MX> &acc0, vector<vector<MX>> &inputs){
    vector<MX> acc = acc0;
    vector<vector<MX>> outputs;
    for (int k=0; k<inputs.size(); k++){
        vector<MX> local_inputs = {acc, inputs[k]}; // psuedo-code
        vector<MX> local_outputs = this.call(local_inputs);
        acc = local_outputs[0];
        outputs.push_back(local_outputs[1:]); // psuedo-code
    }
    return {outputs, acc}; // psuedo-code
}
```

**Comment by jaeandersson** [comment]
fold just a special case of map acc (no local outputs). map acc has simple ad rules.

**Comment by ghorn** [comment]
fold is not a special case of map. Please see my reference implementation above

**Comment by ghorn** [comment]
oh mapAccum, sorry you're right. Map and fold are both special cases of mapAccum

**Comment by jaeandersson** [comment]
map and mapAccum have pretty simple derivative rules. fold doesn't, which is why i'm not sure the operation makes sense. It also makes sense to have a tail recursion operation (a for/while loop). I don't think it can be replaced with any of the operations discussed so far.

**Comment by ghorn** [comment]
Are you sure mapAccum has a simple derivative rule? Fold is a special case of mapAccum where you discard the stage-wise outputs and only return the accumulator. If mapAccum has a simple derivative rule, then fold should also

**Comment by jaeandersson** [comment]
mapAccum (without loss of generality):
```
inputs: x0, u0,u1, ..., u_{N-1}
for i=0, ..., N-1
  y_i = f(x_i, u_i)
  x_{i+1} = f(x_i, u_i)
outputs: x_N, y0, y1, ..., y_{n-1}
```

**Comment by jaeandersson** [comment]
The reverse mode AD can be calculated using two mapAccum.
1. Taping mapAccum (with g(x,u) := x)
```
inputs: x[0], u[0],..., u[N-1]
for k=0, ..., N-1
  x[k+1] = f(x[k], u[k])
outputs: x[1], ..., x[N]
```
2. Reverse sweep mapAccum:
```
inputs: z[0], v[0],..., v[N-1]
for k=0, ..., N-1
  w[k] = adj_g(z[k], v[k])
  z[k+1] = adj_f(z[k], v[k])
outputs: w[0], ..., w[N-1], z[N]
```
Where:
  * `v[k] := {u[N-1-k], x[N-k]}`
  * `w[k] := adj_u[N-1-k]`
  * `z[k] := adj_x[N-k]`

**Comment by jaeandersson** [comment]
> If mapAccum has a simple derivative rule, then fold should also

No. Unless you are able to invert the equation `x[k+1] = f(x[k], u[k])`.

**Comment by ghorn** [comment]
Do you agree that `fold` is a special case of `mapAccum`?

**Comment by jaeandersson** [comment]
It is. But you can't add fold without first adding mapaccum. If you derive a reverse mode derivative rule for fold, you'll see that you'll need mapaccum to express it. So mapaccum is the natural operation to add, and fold isn't really needed. i should update the issue title.

**Comment by ghorn** [comment]
Gotcha, makes sense. I would still expose fold in the api

**Comment by jaeandersson** [comment]
Sure. Function::fold still makes sense, just not the node.

**Comment by jaeandersson** [comment]
Both fold and mapaccum nodes also possibility. Difficult to say before starting to code.

**Comment by jgillis** [comment]
Is there a need to make f and g explicit, I wonder?
What if you use just one F, with first output of F the accumulators outcome and the remainders y's?

**Comment by jgillis** [comment]
This is essentially implemented. Propose spawning a new ticket to discuss potential missing features. e.g. #1531
